src/twisted/test/test_defer.py

```
@@ -11,8 +11,17 @@
 import gc, traceback
 import re
 
+from twisted.python.compat import _PY3, _PY34PLUS
+
+if _PY34PLUS:
+    from asyncio import new_event_loop, Future, CancelledError
+    asyncSkip = None
+else:
+    asyncSkip = "asyncio not available before python 3.4"
+
```

It would be more robust here, perhaps, to catch ImportError.  As it is, I'm assuming you checked about which versions include the asyncio module, but maybe a backport of it comes along…

Anyway, seems better to test for the feature directly if you can, rather than testing for interpreter versions, and here you should be able to test for the feature.

---

I've gone back and forth over this throughout the years, so I'm sympathetic to that point of view, but my current thinking goes like this:

We know that `asyncio` is part of the stdlib on 3.4+.  It _must_ be importable there or there's some kind of major problem with the python installation.  In the case where we expect it to be available because the interpreter says it should be, we go ahead and import it, and fail hard if something is so messed up that it's not importable.  In the case where we _don't_ think it should be available, there's no point in trying.

For _application_ code there's a benefit to being lenient and trying to function in the face of some ambiguity, but for _tests_ this still strikes me as the better way of dealing with conditional imports.

I'm open to being convinced otherwise though.

---

Hrm, that's fair.  I tend to think that it's better (and the intentions often clearer; though in this case the strings are pretty clear, so it's a thin argument here) to test for the think you need directly (import THING), and not indirectly (if python version is X then assume THING).